File: tests/benchmarks/samples/cat-go/001_cart_service.go
// Package cart implements shopping cart item management, discount
// application, and checkout total calculation for the storefront backend.
package cart

import (
	"errors"
	"fmt"
)

// MaxQuantityPerItem caps how many units of a single SKU may be added to a cart.
const MaxQuantityPerItem = 20

// FreeShippingThresholdCents is the subtotal, in cents, at or above which
// shipping is free.
const FreeShippingThresholdCents = 5000

// ErrQuantityExceeded is returned when a requested quantity exceeds
// MaxQuantityPerItem.
var ErrQuantityExceeded = errors.New("quantity exceeds per-item limit")

// LineItem represents one SKU and its quantity/unit price in a cart.
type LineItem struct {
	SKU            string
	Quantity       int
	UnitPriceCents int
}

// Storage persists a cart's line items.
type Storage interface {
	Persist(items map[string]LineItem) error
}

// CartService manages the line items for a single customer's cart.
type CartService struct {
	storage Storage
	items   map[string]LineItem
}

// NewCartService constructs a CartService backed by the given storage.
func NewCartService(storage Storage) *CartService {
	return &CartService{storage: storage, items: make(map[string]LineItem)}
}

// AddItem adds quantity units of sku to the cart, merging with any existing
// quantity for the same SKU.
func (c *CartService) AddItem(sku string, quantity int, unitPriceCents int) error {
	if quantity > MaxQuantityPerItem {
		return fmt.Errorf("%w: %d", ErrQuantityExceeded, quantity)
	}
	existing, ok := c.items[sku]
	next := quantity
	if ok {
		next += existing.Quantity
	}
	c.items[sku] = LineItem{SKU: sku, Quantity: next, UnitPriceCents: unitPriceCents}
	return c.storage.Persist(c.items)
}

// RemoveItem removes sku from the cart entirely.
func (c *CartService) RemoveItem(sku string) error {
	delete(c.items, sku)
	return c.storage.Persist(c.items)
}

// SubtotalCents returns the sum of quantity times unit price across all
// items currently in the cart.
func (c *CartService) SubtotalCents() int {
	total := 0
	for _, item := range c.items {
		total += item.Quantity * item.UnitPriceCents
	}
	return total
}

// ShippingCents returns the shipping cost in cents, which is zero once the
// subtotal reaches FreeShippingThresholdCents.
func (c *CartService) ShippingCents() int {
	if c.SubtotalCents() >= FreeShippingThresholdCents {
		return 0
	}
	return 599
}

File: tests/benchmarks/samples/cat-rust/001_cart_service.rs
//! Shopping cart service — item management, discount codes, and checkout
//! total calculation for the storefront backend.

use std::collections::HashMap;

/// Caps how many units of a single SKU may be added to a cart.
pub const MAX_QUANTITY_PER_ITEM: u32 = 20;

/// The subtotal, in cents, at or above which shipping is free.
pub const FREE_SHIPPING_THRESHOLD_CENTS: u64 = 5000;

/// One SKU and its quantity/unit price in a cart.
#[derive(Debug, Clone)]
pub struct LineItem {
    pub sku: String,
    pub quantity: u32,
    pub unit_price_cents: u64,
}

/// Errors that can occur while mutating a cart.
#[derive(Debug)]
pub enum CartError {
    QuantityExceeded(u32),
    ItemNotFound(String),
}

/// Manages the line items for a single customer's cart.
pub struct CartService {
    items: HashMap<String, LineItem>,
}

impl CartService {
    /// Construct an empty cart.
    pub fn new() -> Self {
        CartService {
            items: HashMap::new(),
        }
    }

    /// Add `quantity` units of `sku` to the cart, merging with any existing
    /// quantity for the same SKU.
    pub fn add_item(&mut self, sku: &str, quantity: u32, unit_price_cents: u64) -> Result<(), CartError> {
        if quantity > MAX_QUANTITY_PER_ITEM {
            return Err(CartError::QuantityExceeded(quantity));
        }
        let next_quantity = match self.items.get(sku) {
            Some(existing) => existing.quantity + quantity,
            None => quantity,
        };
        self.items.insert(
            sku.to_string(),
            LineItem {
                sku: sku.to_string(),
                quantity: next_quantity,
                unit_price_cents,
            },
        );
        Ok(())
    }

    /// Remove `sku` from the cart entirely.
    pub fn remove_item(&mut self, sku: &str) -> Result<(), CartError> {
        match self.items.remove(sku) {
            Some(_) => Ok(()),
            None => Err(CartError::ItemNotFound(sku.to_string())),
        }
    }

    /// Return the sum of quantity times unit price across all items.
    pub fn subtotal_cents(&self) -> u64 {
        self.items
            .values()
            .map(|item| item.quantity as u64 * item.unit_price_cents)
            .sum()
    }

    /// Return the shipping cost in cents, zero once the subtotal reaches
    /// FREE_SHIPPING_THRESHOLD_CENTS.
    pub fn shipping_cents(&self) -> u64 {
        if self.subtotal_cents() >= FREE_SHIPPING_THRESHOLD_CENTS {
            0
        } else {
            599
        }
    }
}
